src/services/refactors/annotateWithTypeFromJSDoc.ts

```diff
@@ -0,0 +1,224 @@
+            && !parameter.dotDotDotToken        // parameter may not be rest
+            && !parameter.questionToken         // parameter may not be optional
+            && !parameter.type                  // parameter may not have a type annotation
+            && !parameter.initializer           // parameter may not have an initializer
+            && isIdentifier(parameter.name);    // parameter name must be identifier
+    }
+
+    function transformJSDocType(node: Node): Node | undefined {
+        if (node === undefined) {
+            return undefined;
+        }
+        switch (node.kind) {
+            case SyntaxKind.JSDocAllType:
+            case SyntaxKind.JSDocUnknownType:
+                return createTypeReferenceNode("any", emptyArray);
+            case SyntaxKind.JSDocOptionalType:
+                return visitJSDocOptionalType(node as JSDocOptionalType);
+            case SyntaxKind.JSDocNonNullableType:
+                return transformJSDocType((node as JSDocNonNullableType).type);
+            case SyntaxKind.JSDocNullableType:
+                return visitJSDocNullableType(node as JSDocNullableType);
+            case SyntaxKind.JSDocVariadicType:
+                return visitJSDocVariadicType(node as JSDocVariadicType);
+            case SyntaxKind.JSDocFunctionType:
+                return visitJSDocFunctionType(node as JSDocFunctionType);
+            case SyntaxKind.Parameter:
+                return visitJSDocParameter(node as ParameterDeclaration);
+            case SyntaxKind.TypeReference:
+                return visitJSDocTypeReference(node as TypeReferenceNode);
+            default:
+                return visitEachChild(node, transformJSDocType, /*context*/ undefined) as TypeNode;
+        }
+    }
+
+    function visitJSDocOptionalType(node: JSDocOptionalType) {
+        return createUnionTypeNode([visitNode(node.type, transformJSDocType), createTypeReferenceNode("undefined", emptyArray)]);
+    }
+
+    function visitJSDocNullableType(node: JSDocNullableType) {
+        return createUnionTypeNode([visitNode(node.type, transformJSDocType), createTypeReferenceNode("null", emptyArray)]);
+    }
+
+    function visitJSDocVariadicType(node: JSDocVariadicType) {
+        return createArrayTypeNode(visitNode(node.type, transformJSDocType));
+    }
+
+    function visitJSDocFunctionType(node: JSDocFunctionType) {
+        const parameters = node.parameters && node.parameters.map(transformJSDocType);
+        return createFunctionTypeNode(emptyArray, parameters as ParameterDeclaration[], node.type);
+    }
+
+    function visitJSDocParameter(node: ParameterDeclaration) {
+        const name = node.name || "arg" + node.parent.parameters.indexOf(node);
+        return createParameter(node.decorators, node.modifiers, node.dotDotDotToken, name, node.questionToken, node.type, node.initializer);
+    }
+
+    function visitJSDocTypeReference(node: TypeReferenceNode) {

```

Comment:
Would it be worth converting `Array.<T>` to `(T)[]`, rather than `Array<T>`?

Reply:
Why parentheses?

Reply:
@weswigham Nope. I don't want to introduce arbitrary style fixups, just ones that we believe are nearly always wrong. And `Array<T>` is a style choice that is favoured by lots (a majority?) of the JS community.

Reply:
@DanielRosenwasser I just included those in case T was `string | number`, to serve as a reminder parens may be needed. But if @sandersn wants to preserve the existing style, that's fine.